src/spyglass/common/common_optogenetics.py

```diff
@@ -0,0 +1,377 @@
+        # insert keys
+        self.insert(epoch_inserts)
+        self.RippleTrigger.insert(ripple_inserts)
+        self.ThetaTrigger.insert(theta_inserts)
+        self.SpeedConditional.insert(speed_inserts)
+        self.SpatialConditional.insert(spatial_inserts)
+
+    def get_stimulus_on_intervals(self, key):
+        key = (self & key).fetch1("KEY")
+        nwb = (self & key).fetch_nwb()[0]
+        stimulus = nwb["stimulus"]
+        stim_time = stimulus.timestamps
+        stim_data = stimulus.data
+        # restrict data to the epoch
+        epoch_interval = (IntervalList & (TaskEpoch & key)).fetch_interval()
+        epoch_ind = epoch_interval.contains(stim_time, as_indices=True)
+        stim_time = stim_time[epoch_ind]
+        stim_data = stim_data[epoch_ind]
+        # make interval for stimulu on times
+        t_on = stim_time[stim_data == 1]
+        t_off = stim_time[stim_data == 0]
+        if t_off[0] < t_on[0]:
+            t_off = t_off[1:]
+        if t_on[-1] > t_off[-1]:
+            t_off = np.append(t_off, stim_time[-1])
+        stim_on_interval = np.array([t_on, t_off]).T
+        return stim_on_interval
+
+    class RippleTrigger(SpyglassMixin, dj.Part):
+        definition = """
+        # Ripple trigger for optogenetic stimulation
+        -> master
+        ---
+        threshold_sd: float  # standard deviation threshold for ripple detection
+        n_above_threshold: int  # number of samples above threshold for ripple detection
+        lockout_period: float  # lockout period in sample steps
+        """
+
+    class ThetaTrigger(SpyglassMixin, dj.Part):
+        definition = """
+        # Theta trigger for optogenetic stimulation
+        -> master
+        ---
+        filter_phase: float # target phase of the trigger
+        reference_ntrode: int # reference ntrode for the trigger
+        lockout_period: float # lockout period in sample steps
+        """
+
+    class SpeedConditional(SpyglassMixin, dj.Part):
+        definition = """
+        # Speed conditional for optogenetic stimulation
+        -> master
+        ---
+        speed_threshold: float # speed threshold for optogenetic stimulation (cm/s)
+        active_above_threshold: bool # whether the stimulation is active above or below the threshold
+        """
+
+    class SpatialConditional(SpyglassMixin, dj.Part):
+        definition = """
+        # Spatial conditional for optogenetic stimulation
+        -> master
+        ---
+        nodes: longblob # list of nodes defining polygonal area for optogenetic stimulation
+        """
+
+    def get_stimulus_timeseries(self):
+        if not len(self) == 1:
+            raise ValueError(
+                "get_stimulus_timeseries() only works for single entries"
+            )
+        return self.fetch_nwb()[0]["stimulus"]
+
+
+@schema
+class Virus(SpyglassMixin, dj.Manual):
+    definition = """
+    # Virus information
+    virus_name: varchar(255)
```

Comment:
That's a large primary key (#630 ) - do you expect this table to ever be fk-ref'd?